File: app/src/main/java/com/sofps/inspirationalquotes/data/QuoteApi.java
package com.sofps.inspirationalquotes.data;

import com.google.gson.annotations.SerializedName;
import java.util.List;

// TODO this data model is incorrect and it's not parsed correctly when trying to get the quote of the day from the api
public class QuoteApi {

    @SerializedName("author") private String mAuthor;
    @SerializedName("background") private String mBackground;
    @SerializedName("category") private String mCategory;
    @SerializedName("date") private String mDate;
    @SerializedName("id") private String mId;
    @SerializedName("length") private Object mLength;
    @SerializedName("permalink") private String mPermalink;
    @SerializedName("quote") private String mQuote;
    @SerializedName("tags") private List<String> mTags;
    @SerializedName("title") private String mTitle;

    public String getAuthor() {
        return mAuthor;
    }

    public String getBackground() {
        return mBackground;
    }

    public String getCategory() {
        return mCategory;
    }

    public String getDate() {
        return mDate;
    }

    public String getId() {
        return mId;
    }

    public Object getLength() {
        return mLength;
    }

    public String getPermalink() {
        return mPermalink;
    }

    public String getQuote() {
        return mQuote;
    }

    public List<String> getTags() {
        return mTags;
    }

    public String getTitle() {
        return mTitle;
    }

    public static class Builder {

        private String mAuthor;
        private String mBackground;
        private String mCategory;
        private String mDate;
        private String mId;
        private Object mLength;
        private String mPermalink;
        private String mQuote;
        private List<String> mTags;
        private String mTitle;

        public QuoteApi.Builder withAuthor(String author) {
            mAuthor = author;
            return this;
        }

        public QuoteApi.Builder withBackground(String background) {
            mBackground = background;
            return this;
        }

        public QuoteApi.Builder withCategory(String category) {
            mCategory = category;
            return this;
        }

        public QuoteApi.Builder withDate(String date) {
            mDate = date;
            return this;
        }

        public QuoteApi.Builder withId(String id) {
            mId = id;
            return this;
        }

        public QuoteApi.Builder withLength(Object length) {
            mLength = length;
            return this;
        }

        public QuoteApi.Builder withPermalink(String permalink) {
            mPermalink = permalink;
            return this;
        }

        public QuoteApi.Builder withQuote(String quote) {
            mQuote = quote;
            return this;
        }

        public QuoteApi.Builder withTags(List<String> tags) {
            mTags = tags;
            return this;
        }

        public QuoteApi.Builder withTitle(String title) {
            mTitle = title;
            return this;
        }

        public QuoteApi build() {
            QuoteApi QuoteDao = new QuoteApi();
            QuoteDao.mAuthor = mAuthor;
            QuoteDao.mBackground = mBackground;
            QuoteDao.mCategory = mCategory;
            QuoteDao.mDate = mDate;
            QuoteDao.mId = mId;
            QuoteDao.mLength = mLength;
            QuoteDao.mPermalink = mPermalink;
            QuoteDao.mQuote = mQuote;
            QuoteDao.mTags = mTags;
            QuoteDao.mTitle = mTitle;
            return QuoteDao;
        }
    }
}

File: app/src/main/java/com/sofps/inspirationalquotes/data/QuoteApi.kt
package com.sofps.inspirationalquotes.data

import com.google.gson.annotations.SerializedName

// TODO this data model is incorrect and it's not parsed correctly when trying to get the quote of the day from the api
class QuoteApi {

    @SerializedName("author")
    var author: String? = null
        private set
    @SerializedName("background")
    var background: String? = null
        private set
    @SerializedName("category")
    var category: String? = null
        private set
    @SerializedName("date")
    var date: String? = null
        private set
    @SerializedName("id")
    var id: String? = null
        private set
    @SerializedName("length")
    var length: Any? = null
        private set
    @SerializedName("permalink")
    var permalink: String? = null
        private set
    @SerializedName("quote")
    var quote: String? = null
        private set
    @SerializedName("tags")
    var tags: List<String>? = null
        private set
    @SerializedName("title")
    var title: String? = null
        private set

    val isValid: Boolean
        get() = author != null && author!!.isNotEmpty() && quote != null && quote!!.isNotEmpty()

    class Builder {

        private var author: String? = null
        private var background: String? = null
        private var category: String? = null
        private var date: String? = null
        private var id: String? = null
        private var length: Any? = null
        private var permalink: String? = null
        private var quote: String? = null
        private var tags: List<String>? = null
        private var title: String? = null

        fun withAuthor(author: String): Builder {
            this.author = author
            return this
        }

        fun withBackground(background: String): Builder {
            this.background = background
            return this
        }

        fun withCategory(category: String): Builder {
            this.category = category
            return this
        }

        fun withDate(date: String): Builder {
            this.date = date
            return this
        }

        fun withId(id: String): Builder {
            this.id = id
            return this
        }

        fun withLength(length: Any): Builder {
            this.length = length
            return this
        }

        fun withPermalink(permalink: String): Builder {
            this.permalink = permalink
            return this
        }

        fun withQuote(quote: String): Builder {
            this.quote = quote
            return this
        }

        fun withTags(tags: List<String>): Builder {
            this.tags = tags
            return this
        }

        fun withTitle(title: String): Builder {
            this.title = title
            return this
        }

        fun build() =
                QuoteApi().apply {
                    this.author = author
                    this.background = background
                    this.category = category
                    this.date = date
                    this.id = id
                    this.length = length
                    this.permalink = permalink
                    this.quote = quote
                    this.tags = tags
                    this.title = title
                }
    }
}
